### PR TITLE
Replace typecheck with beartype

### PR DESCRIPTION
**Pull request by davidt0x**
This is a big PR to remove `typecheck-decorator` dependency and replace with [beartype](https://github.com/beartype/beartype). This involved the following changes:

- Replace all non PEP compliant type hints and references to typecheck module with standard PEP compliant type hint equivalents where possible.
- Replace all typecheck decorations on functions with beartype decorations to enable runtime checking similar to typecheck-decorator
- Created psyneulink._typing module that contains type aliases for supporting Python 3.7 and 3.8.

Some of the allowed types for parameters are very complex. For instance, the allowed types for parameter specifications, see [here](https://github.com/PrincetonUniversity/PsyNeuLink/blob/b6bb8a1dccbb9b4219423ed9947236a50e968e07/psyneulink/core/globals/utilities.py#L316). This leads to some pretty verbose error messages when parameters are invalid, for example:

```
pnl.Logistic(gain='test')
Traceback (most recent call last):
  File "C:\Anaconda3\envs\pnl_py37\lib\site-packages\IPython\core\interactiveshell.py", line 3524, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-47e065d9467d>", line 1, in <module>
    pnl.Logistic(gain='test')
  File "<@beartype(psyneulink.core.components.functions.nonstateful.transferfunctions.Logistic.__init__) at 0x1ef8a1dedc8>", line 146, in __init__
  File "C:\Anaconda3\envs\pnl_py37\lib\site-packages\beartype\_decor\_error\errormain.py", line 302, in raise_pep_call_exception
    f'{exception_prefix}violates type hint {repr(hint)}, as '
beartype.roar.BeartypeCallHintParamViolation: @beartyped Logistic.__init__() parameter gain='test' violates type hint typing.Union[numbers.Number, numpy.ndarray, list, tuple, dict, function, ForwardRef('psyneulink.core.components.shellclasses.Projection'), ForwardRef('psyneulink.core.components.mechanisms.ControlMechanism'), typing.Type[ForwardRef('psyneulink.core.components.mechanisms.ControlMechanism')], ForwardRef('psyneulink.core.components.projections.ControlProjection'), typing.Type[ForwardRef('psyneulink.core.components.projections.ControlProjection')], ForwardRef('psyneulink.core.components.ports.ControlSignal'), typing.Type[ForwardRef('psyneulink.core.components.ports.ControlSignal')], ForwardRef('psyneulink.core.components.mechanisms.GatingMechanism'), typing.Type[ForwardRef('psyneulink.core.components.mechanisms.GatingMechanism')], ForwardRef('psyneulink.core.components.projections.GatingProjection'), typing.Type[ForwardRef('psyneulink.core.components.projections.GatingProjection')], ForwardRef('psyneulink.core.components.ports.GatingSignal'), typing.Type[ForwardRef('psyneulink.core.components.ports.GatingSignal')], ForwardRef('psyneulink.core.components.mechanisms.LearningMechanism'), typing.Type[ForwardRef('psyneulink.core.components.mechanisms.LearningMechanism')], ForwardRef('psyneulink.core.components.projections.LearningProjection'), typing.Type[ForwardRef('psyneulink.core.components.projections.LearningProjection')], ForwardRef('psyneulink.core.components.ports.LearningSignal'), typing.Type[ForwardRef('psyneulink.core.components.ports.LearningSignal')], ForwardRef('psyneulink.library.components.projections.AutoAssociativeProjection'), typing.Type[ForwardRef('psyneulink.library.components.projections.AutoAssociativeProjection')], ForwardRef('psyneulink.core.components.projections.MappingProjection'), typing.Type[ForwardRef('psyneulink.core.components.projections.MappingProjection')], ForwardRef('psyneulink.library.components.projections.MaskedMappingProjection'), typing.Type[ForwardRef('psyneulink.library.components.projections.MaskedMappingProjection')], typing_extensions.Literal['LEARNING', 'bias', 'control', 'gain', 'gate', 'leak', 'offset', 'ControlSignal', 'ControlProjection'], NoneType], as 'test':
* Not tuple, <class "numpy.ndarray">, <class "builtins.function">, dict, <class "builtins.NoneType">, list, type, or <class "numbers.Number">.
* 'test' not instance of <class "psyneulink.core.components.shellclasses.Projection">.
* 'test' not instance of <class "psyneulink.core.components.mechanisms.modulatory.control.controlmechanism.ControlMechanism">.
* 'test' not instance of <class "psyneulink.core.components.projections.modulatory.controlprojection.ControlProjection">.
* 'test' not instance of <class "psyneulink.core.components.ports.modulatorysignals.controlsignal.ControlSignal">.
* 'test' not instance of <class "psyneulink.core.components.mechanisms.modulatory.control.gating.gatingmechanism.GatingMechanism">.
* 'test' not instance of <class "psyneulink.core.components.projections.modulatory.gatingprojection.GatingProjection">.
* 'test' not instance of <class "psyneulink.core.components.ports.modulatorysignals.gatingsignal.GatingSignal">.
* 'test' not instance of <class "psyneulink.core.components.mechanisms.modulatory.learning.learningmechanism.LearningMechanism">.
* 'test' not instance of <class "psyneulink.core.components.projections.modulatory.learningprojection.LearningProjection">.
* 'test' not instance of <class "psyneulink.core.components.ports.modulatorysignals.learningsignal.LearningSignal">.
* 'test' not instance of <class "psyneulink.library.components.projections.pathway.autoassociativeprojection.AutoAssociativeProjection">.
* 'test' not instance of <class "psyneulink.core.components.projections.pathway.mappingprojection.MappingProjection">.
* 'test' not instance of <class "psyneulink.library.components.projections.pathway.maskedmappingprojection.MaskedMappingProjection">.
* 'test' != 'LEARNING', 'bias', 'control', 'gain', 'gate', 'leak', 'offset', 'ControlSignal', or 'ControlProjection'.
```

We can override this messaging of course by creating our own decorator that catches these exceptions. However, it might be better to constrain the type annotations more carefully on a per parameter basis.

I also tried testing on python 3.10 since typecheck decorator was holding us up there but we still have some other dependencies that seem to be blocking us there. I was getting some problems with modeci_mdf which I will need to look into.

### CODE REVIEW

File: psyneulink/_typing.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    Union as Union,
+    Type as Type,
+    Callable as Callable,
+    Optional as Optional,
+    Any as Any,
+    Tuple as Tuple,
+    List as List,
+    Dict as Dict,
+    Iterable as Iterable,
+    Set as Set,
```
Comment (jvesely) [inline psyneulink/_typing.py:18]: Why can't these be just imported from typing?

File: .github/workflows/pnl-ci.yml [hunk 2]
```diff
@@ -17,26 +17,26 @@ jobs:
...
             os: windows-latest
           # code-coverage build on macos python 3.9
-          - python-version: 3.9
+          - python-version: '3.9'
             os: macos-latest
             extra-args: '--cov=psyneulink'
         exclude:
           # 3.7 is broken on macos-11, https://github.com/actions/virtual-environments/issues/4230
-          - python-version: 3.7
+          - python-version: '3.7'
```
Comment (jvesely) [inline .github/workflows/pnl-ci.yml:39]: please don't bundle this with unrelated changes.

File: psyneulink/core/components/functions/nonstateful/combinationfunctions.py [hunk 3]
```diff
@@ -723,19 +725,25 @@ class Parameters(CombinationFunction.Parameters):
...
-                 # exponents: tc.optional(parameter_spec)=None,
+                 # weights:  Optional[ValidParamSpecType] = None,
+                 # exponents:  Optional[ValidParamSpecType] = None,
                  weights=None,
                  exponents=None,
                  default_variable=None,
-                 operation: tc.optional(tc.enum(SUM, PRODUCT)) = None,
-                 scale: tc.optional(parameter_spec) = None,
-                 offset: tc.optional(parameter_spec) = None,
+                 operation: Optional[Literal['sum', 'product']] = None,
```
Comment (jvesely) [inline psyneulink/core/components/functions/nonstateful/combinationfunctions.py:735]: Does this work with `kw.SUM`/`kw.PRODUCT` ?

File: psyneulink/_typing.py (hunk 1 above)
Comment (davidt0x) [reply psyneulink/_typing.py:18]: See https://github.com/beartype/beartype#pep-585-deprecations. This silences a warning from beartype. Apparently most of typing is slated for deprecation in a few years. Things like List and Tuple can be replaced with list and tuple in python 3.9 but since we are supporting <3.9 we can't use these. I think long term we can replace these when we drop support for 3.8.

File: psyneulink/core/components/functions/nonstateful/combinationfunctions.py (hunk 3 above)
Comment (davidt0x) [reply psyneulink/core/components/functions/nonstateful/combinationfunctions.py:735]: Do you mean indexing typing.Literal with kw.SUM? Unfortunately no, since they are not defined statically. If we made all keywords enums then you CAN index Literal with them it seems. I was thinking of doing this but I couldn't come up with something that didn't introduce a ton of code changes. They probably should be enums though. One way that might work would be to move them all to enums and then programmatically instantiate the keyword globals from the enums.

File: psyneulink/_typing.py (hunk 1 above)
Comment (jvesely) [reply psyneulink/_typing.py:18]: so the code now uses `_typing` instead of `beartype.typing`?
Can't the same be achieved by just `import beartype.typing as typing` in pnl `__init__.py` ?
if not can all custom PNL types be moved in this file?

File: psyneulink/core/components/functions/nonstateful/combinationfunctions.py (hunk 3 above)
Comment (jvesely) [reply psyneulink/core/components/functions/nonstateful/combinationfunctions.py:735]: yeah, having keywords as enums would be nice, but that's not a good fit for this change.
would it work if keywords were typed hinted at definition, something like `SUM:Literal['sum'] = 'sum'`?
are there many keywords that would need such treatment?

File: psyneulink/_typing.py (hunk 1 above)
Comment (davidt0x) [reply psyneulink/_typing.py:18]: Yeah, the _typing.py I was thinking could be a place to consolidate typing stuff down the line. I created it mainly for the typing_extensions import of Literal for 3.7 actually. But I think we could move other stuff there as well.

File: .github/workflows/pnl-ci.yml (hunk 2 above)
Comment (davidt0x) [reply .github/workflows/pnl-ci.yml:39]: Yeah sure, I can revert this completely.

Comment (davidt0x) [reply .github/workflows/pnl-ci.yml:39]: Yeah, I can revert this completely.

File: psyneulink/core/components/functions/nonstateful/combinationfunctions.py (hunk 3 above)
Comment (davidt0x) [reply psyneulink/core/components/functions/nonstateful/combinationfunctions.py:735]: ok @jvesely and @jdcpni,

something like this seems to work:

```
from beartype import beartype
from typing_extensions import Literal, Final
from enum import Enum


class Keyword(Enum):
    SUM = "sum"
    PRODUCT = "product"


SUM: Final = Keyword.SUM
PRODUCT: Final = Keyword.PRODUCT


@beartype
def test_literal(b: Literal[Keyword.SUM, Keyword.PRODUCT]) -> None:
    print(b)


# Passes runtime and static type checks.
test_literal(SUM)
test_literal(PRODUCT)
test_literal(Keyword.SUM)
test_literal(Keyword.PRODUCT)

# Fails runtime type check
test_literal("sum")
```

This would require having two lines in keywords.py for each keyword. One for the enum, and one for the module constant variable for that keyword. So if someone wants to add a keyword, they would need to know about this and put it in both places. If this looks good, I can try to implement it like this. I anticipate the there are some places in the code that expect these keywords to actually be strings which is no longer the case using this approach. This can probably be fixed by simply using `SUM.value` in those places though.

What do you both think?

**Comment by jvesely** [comment]
can you add:
```
Fixes: https://github.com/PrincetonUniversity/PsyNeuLink/issues/1703
```
to merge commit description? it will help with issue tracking.

another nit: I'd be nice to remove the 33 newly introduced redundant imports (see the comment from lgtm)

**Comment by davidt0x** [comment]
@jdcpni and @kmantel, can this be merged now? I know it is not perfect in regards to the verbosity of error messages but I think these could be worked out in a separate PR. I also did a test merge with the allclose-changes and that didn't seem to have any conflicts.

**Comment by jdcpni** [comment]
Good by me.


jdc



On Apr 5, 2023, at 4:49 PM, David Turner ***@***.***> wrote:



@jdcpni<https://github.com/jdcpni> and @kmantel<https://github.com/kmantel>, can this be merged now? I know it is not perfect in regards to the verbosity of error messages but I think these could be worked out in a separate PR. I also did a test merge with the allclose-changes and that didn't seem to have any conflicts.

File: psyneulink/core/components/functions/nonstateful/combinationfunctions.py [hunk 4]
```diff
@@ -727,19 +728,25 @@ class Parameters(CombinationFunction.Parameters):
...
+                 scale: Optional[ValidParamSpecType] = None,
+                 offset: Optional[ValidParamSpecType] = None,
                  params=None,
                  owner=None,
-                 prefs: tc.optional(is_pref_set) = None):
+                 prefs:  Optional[ValidPrefSet] = None):
+
+        if scale is not None and not np.isscalar(scale):
+            raise ValueError("scale must be a scalar")
```
Comment (kmantel) [inline psyneulink/core/components/functions/nonstateful/combinationfunctions.py:746]: If desired, this can be replaced with a method like this (name important) on `Reduce.Parameters`:

```python
        def _validate_scale(self, scale):
            if not np.isscalar(scale):
                return "scale must be a scalar"
```
which would also support subclasses and validation at post-construction assignments

File: psyneulink/core/components/functions/stateful/integratorfunctions.py [hunk 5]
```diff
@@ -2431,24 +2433,33 @@ def _parse_initializer(self, initializer):
...
 
+        # Make sure rate is a 1D array or scalar
+        if rate is not None:
+            if isinstance(rate, np.ndarray):
+                if rate.ndim > 1:
+                    raise ValueError(f"incompatible value ({rate}) for rate parameter, must be 1D array or scalar")
+            else:
+                if not np.isscalar(rate) and type(rate) != list:
+                    raise ValueError(f"incompatible value ({rate}) for rate parameter, must be 1D array or scalar")
+
```
Comment (kmantel) [inline psyneulink/core/components/functions/stateful/integratorfunctions.py:2462]: Same suggestion to write a `_validate_rate` function

**Comment by kmantel** [comment]
> @jdcpni and @kmantel, can this be merged now? I know it is not perfect in regards to the verbosity of error messages but I think these could be worked out in a separate PR. I also did a test merge with the allclose-changes and that didn't seem to have any conflicts.

Sure, in the interest of supporting 3.10, I think it's ok. The error messages aren't ideal, yes, and more specificity on types required is probably good, but that's a separate task. With that, I think more use of the `Parameter.valid_types` argument could be helpful for some of the typechecks that are basically simple isinstance
